File: samkhya-gpudb/scripts/llm_dummy_backend.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: Prateek Singh
"""Standalone dummy LLM backend for transport-floor measurement.

Used by the file 19 transport-floor cell so the LLM-pluggable corrector
backend can be exercised end-to-end without an API key. The server in
``llm_infer_server.py`` already supports a ``dummy`` backend internally;
this file is a *standalone* HTTP shim that speaks the same wire contract
without depending on FastAPI, useful when a reviewer wants to reproduce
the floor with only the stdlib.

Wire contract (matches ``samkhya-core::residual::llm``)::

    POST /infer  Content-Type: application/json
    { "features": [...], "baseline_estimate": <u64> }
    -> 200 OK
    { "estimate": <u64> }

Strategy
--------

The dummy returns the baseline estimate unchanged. This is the *neutral*
estimate from the cardinality-corrector contract's perspective — no
information is added by the corrector, so the engine behaves as if no
corrector were installed. The cost we measure with the dummy is the
*HTTP transport floor* of the LLM backend: serialize features → POST →
deserialize response. Real LLM backends stack on top of this floor.

Run
---

    python3 samkhya-gpudb/scripts/llm_dummy_backend.py --port 8766

then drive it with the same Rust client (``llm_latency`` binary) that
hits the production server.
"""

from __future__ import annotations

import argparse
import json
import os
import sys
from http.server import BaseHTTPRequestHandler, HTTPServer
from typing import Optional

FEATURE_LEN = 7
U64_MAX = (1 << 64) - 1

# SECURITY-REVIEW-2026-05-17.md (H4 + M1).
BODY_MAX_BYTES = 8 * 1024 * 1024
MAX_INFER_BATCHES = 1024


class Handler(BaseHTTPRequestHandler):
    """One-shot HTTP handler — minimal correctness, maximal stability."""

    server_version = "samkhya-llm-dummy/1.0"

    def _read_request_body(self) -> Optional[bytes]:
        """Read a bounded Content-Length or chunked request body."""
        transfer_encoding = self.headers.get("Transfer-Encoding", "").lower()
        if transfer_encoding:
            if transfer_encoding != "chunked":
                self.send_error(400, "unsupported Transfer-Encoding")
                return None

            body = bytearray()
            overflow = False
            while True:
                size_line = self.rfile.readline(128)
                if (
                    not size_line
                    or len(size_line) >= 128
                    or not size_line.endswith(b"\r\n")
                ):
                    self.send_error(400, "invalid chunk size")
                    return None
                size_token = size_line[:-2].split(b";", 1)[0]
                if not size_token or any(
                    byte not in b"0123456789abcdefABCDEF" for byte in size_token
                ):
                    self.send_error(400, "invalid chunk size")
                    return None
                chunk_size = int(size_token, 16)
                if chunk_size == 0:
                    # Consume a bounded trailer section through its blank line.
                    for _ in range(32):
                        trailer = self.rfile.readline(8192)
                        if trailer == b"\r\n":
                            break
                        if not trailer or len(trailer) >= 8192:
                            self.send_error(400, "invalid chunk trailer")
                            return None
                    else:
                        self.send_error(400, "too many chunk trailers")
                        return None
                    break

                remaining = chunk_size
                while remaining:
                    block = self.rfile.read(min(remaining, 64 * 1024))
                    if not block:
                        self.send_error(400, "truncated chunk")
                        return None
                    remaining -= len(block)
                    if not overflow:
                        if len(body) + len(block) > BODY_MAX_BYTES:
                            body.clear()
                            overflow = True
                        else:
                            body.extend(block)
                if self.rfile.read(2) != b"\r\n":
                    self.send_error(400, "invalid chunk terminator")
                    return None

            if overflow:
                self.send_error(413, f"request body exceeds {BODY_MAX_BYTES}")
                return None
            return bytes(body)

        try:
            length = int(self.headers.get("Content-Length", "0"))
        except ValueError:
            self.send_error(400, "invalid Content-Length")
            return None
        if length < 0:
            self.send_error(400, "invalid Content-Length")
            return None
        if length > BODY_MAX_BYTES:
            self.send_error(413, f"body {length} exceeds {BODY_MAX_BYTES}")
            return None
        return self.rfile.read(length)

    def do_GET(self) -> None:  # noqa: N802 — BaseHTTPRequestHandler API
        if self.path != "/health":
            self.send_error(404, "not found")
            return
        body = json.dumps(
            {"ok": True, "backend": "dummy", "model": "dummy-echo"}
        ).encode()
        self._respond(200, body)

    def do_POST(self) -> None:  # noqa: N802 — BaseHTTPRequestHandler API
        if self.path != "/infer":
            self.send_error(404, "not found")
            return
        raw = self._read_request_body()
        if raw is None:
            return
        try:
            body = json.loads(raw)
        except Exception as exc:
            # See SECURITY-REVIEW-2026-05-17.md (C3): log full detail to
            # stderr but echo only the exception class on the wire.
            print(f"[llm-dummy] /infer parse err: {exc!r}", file=sys.stderr, flush=True)
            self.send_error(400, f"bad json: {type(exc).__name__}")
            return

        if not isinstance(body, dict):
            self.send_error(400, "json body must be an object")
            return

        features = body.get("features")
        baseline = body.get("baseline_estimate")
        if not isinstance(features, list) or not features:
            self.send_error(400, "missing features")
            return
        if not all(
            not isinstance(value, bool)
            and isinstance(value, (int, float))
            and (not isinstance(value, float) or value == value)
            and value not in (float("inf"), float("-inf"))
            for value in features
        ):
            self.send_error(400, "features must contain only finite numbers")
            return
        if type(baseline) is not int or not 0 <= baseline <= U64_MAX:
            self.send_error(400, "invalid baseline_estimate")
            return
        if len(features) % FEATURE_LEN != 0:
            self.send_error(
                400, f"features length {len(features)} not a multiple of {FEATURE_LEN}"
            )
            return
        batches = len(features) // FEATURE_LEN
        if batches > MAX_INFER_BATCHES:
            self.send_error(413, f"features batch count {batches} exceeds {MAX_INFER_BATCHES}")
            return

        reply = json.dumps({"estimate": baseline}).encode()
        self._respond(200, reply)

    def log_message(self, *_args: object, **_kwargs: object) -> None:
        # Suppress per-request access logs to stdout — we want clean
        # bench output. The Rust client times its own round-trip.
        pass

    def _respond(self, code: int, body: bytes) -> None:
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Connection", "close")
        self.end_headers()
        try:
            self.wfile.write(body)
            self.wfile.flush()
        except BrokenPipeError:
            pass


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--host", default=os.environ.get("SAMKHYA_LLM_HOST", "127.0.0.1")
    )
    parser.add_argument(
        "--port",
        type=int,
        default=int(os.environ.get("SAMKHYA_LLM_PORT", "8766")),
    )
    args = parser.parse_args()
    # SECURITY-REVIEW-2026-05-17.md (H1): the dummy backend is harmless
    # (it doesn't reach upstream LLMs) but the warning still helps the
    # operator notice that an unauthenticated server is exposed.
    if args.host not in ("127.0.0.1", "::1", "localhost"):
        banner = "=" * 70
        print(
            f"\n{banner}\n"
            f"[WARN] samkhya LLM dummy server bound to non-loopback ({args.host}).\n"
            f"[WARN] This server has NO authentication. Ensure network isolation\n"
            f"[WARN] before exposing the address.\n"
            f"{banner}\n",
            file=sys.stderr,
            flush=True,
        )
    print(
        f"[llm-dummy] listening on http://{args.host}:{args.port} "
        f"(GET /health, POST /infer)",
        file=sys.stderr,
        flush=True,
    )
    HTTPServer((args.host, args.port), Handler).serve_forever()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: samkhya-gpudb/scripts/llm_dummy_backend.ts
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Prateek Singh
//
// llm_dummy_backend.ts — zero-dep TypeScript mirror of
// llm_dummy_backend.py. Runs on bare Node 18+; no @anthropic-ai/sdk
// or openai required. Used by the file 19 transport-floor cell so a
// reviewer with only Node installed can replay the round-trip cost
// measurement end-to-end.
//
// Wire contract is identical to llm_infer_server.ts (POST /infer +
// GET /health). The "model" returns baseline_estimate verbatim — the
// number we measure is HTTP round-trip + JSON parse + Node event-loop
// scheduling, nothing else.

import * as http from 'node:http';
import { URL } from 'node:url';

const FEATURE_LEN = 7;
// SECURITY-REVIEW-2026-05-17.md (H4 + M1).
const BODY_MAX_BYTES = 8 * 1024 * 1024;
const MAX_INFER_BATCHES = 1024;

interface DummyState {
  startedAtNs: bigint;
  port: number;
  host: string;
}

const state: DummyState = {
  startedAtNs: 0n,
  port: 8766,
  host: '127.0.0.1',
};

function sendJson(res: http.ServerResponse, status: number, body: unknown): void {
  const buf = Buffer.from(JSON.stringify(body));
  res.writeHead(status, {
    'Content-Type': 'application/json',
    'Content-Length': buf.byteLength,
  });
  res.end(buf);
}

async function readBody(req: http.IncomingMessage, maxBytes: number): Promise<Buffer | null> {
  // SECURITY-REVIEW-2026-05-17.md (H4): same body-size guard as the
  // primary llm_infer_server.ts. Returns null on overflow.
  const declared = req.headers['content-length'];
  if (declared !== undefined) {
    const n = Number(declared);
    if (Number.isFinite(n) && n > maxBytes) return null;
  }
  return new Promise<Buffer | null>((resolve, reject) => {
    const chunks: Buffer[] = [];
    let total = 0;
    let aborted = false;
    req.on('data', (c: Buffer) => {
      if (aborted) return;
      total += c.byteLength;
      if (total > maxBytes) {
        aborted = true;
        req.destroy();
        resolve(null);
        return;
      }
      chunks.push(c);
    });
    req.on('end', () => {
      if (!aborted) resolve(Buffer.concat(chunks));
    });
    req.on('error', reject);
  });
}

function handleHealth(_req: http.IncomingMessage, res: http.ServerResponse): void {
  sendJson(res, 200, {
    ok: true,
    backend: 'dummy',
    model: 'dummy-echo-ts',
    temperature: 0.0,
    max_tokens: 32,
  });
}

async function handleInfer(req: http.IncomingMessage, res: http.ServerResponse): Promise<void> {
  const bodyBytes = await readBody(req, BODY_MAX_BYTES);
  if (bodyBytes === null) {
    sendJson(res, 413, { detail: `request body exceeds ${BODY_MAX_BYTES}` });
    return;
  }
  let body: any;
  try {
    body = JSON.parse(bodyBytes.toString('utf8'));
  } catch (exc) {
    // See SECURITY-REVIEW-2026-05-17.md (C3): log full detail to stderr
    // but echo only the exception class on the wire.
    process.stderr.write(`[llm-dummy] /infer parse err: ${String(exc)}\n`);
    const name = (exc as { constructor?: { name?: string } })?.constructor?.name ?? 'Error';
    sendJson(res, 400, { detail: `invalid json: ${name}` });
    return;
  }
  const features = body.features;
  const baseline = body.baseline_estimate;
  if (!Array.isArray(features) || features.length === 0) {
    sendJson(res, 400, { detail: "missing or empty 'features'" });
    return;
  }
  if (typeof baseline !== 'number' || !Number.isInteger(baseline) || baseline < 0) {
    sendJson(res, 400, { detail: "missing or non-u64 'baseline_estimate'" });
    return;
  }
  if (features.length % FEATURE_LEN !== 0) {
    sendJson(res, 400, {
      detail: `features length ${features.length} not a multiple of ${FEATURE_LEN}`,
    });
    return;
  }
  const batches = features.length / FEATURE_LEN;
  if (batches > MAX_INFER_BATCHES) {
    sendJson(res, 413, {
      detail: `features batch count ${batches} exceeds ${MAX_INFER_BATCHES}`,
    });
    return;
  }
  sendJson(res, 200, { estimate: baseline });
}

function parseArgs(argv: string[]): { host: string; port: number } {
  let host = process.env.SAMKHYA_LLM_HOST ?? '127.0.0.1';
  let port = Number(process.env.SAMKHYA_LLM_PORT ?? '8766');
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--host') host = argv[++i];
    else if (a === '--port') port = Number(argv[++i]);
    else if (a === '-h' || a === '--help') {
      process.stdout.write('usage: node llm_dummy_backend.js [--host H] [--port P]\n');
      process.exit(0);
    } else {
      process.stderr.write(`unknown arg: ${a}\n`);
      process.exit(2);
    }
  }
  return { host, port };
}

const LOOPBACK_HOSTS = new Set(['127.0.0.1', '::1', 'localhost']);

function main(): void {
  const { host, port } = parseArgs(process.argv.slice(2));
  state.host = host;
  state.port = port;
  state.startedAtNs = process.hrtime.bigint();

  // SECURITY-REVIEW-2026-05-17.md (H1): warn on non-loopback bind.
  if (!LOOPBACK_HOSTS.has(host)) {
    const banner = '='.repeat(70);
    process.stderr.write(
      `\n${banner}\n` +
        `[WARN] samkhya LLM dummy server bound to non-loopback (${host}).\n` +
        `[WARN] This server has NO authentication. Ensure network isolation\n` +
        `[WARN] before exposing the address.\n` +
        `${banner}\n`,
    );
  }

  const server = http.createServer((req, res) => {
    const url = new URL(req.url ?? '/', `http://${req.headers.host ?? host}`);
    if (req.method === 'GET' && url.pathname === '/health') {
      handleHealth(req, res);
      return;
    }
    if (req.method === 'POST' && url.pathname === '/infer') {
      handleInfer(req, res).catch((exc) => {
        process.stderr.write(`[llm-dummy] /infer crashed: ${String(exc)}\n`);
        sendJson(res, 500, { detail: 'internal error' });
      });
      return;
    }
    sendJson(res, 404, { detail: 'not found' });
  });

  server.listen(port, host, () => {
    process.stderr.write(`[llm-dummy] listening on http://${host}:${port}\n`);
  });

  const shutdown = (sig: string) => {
    process.stderr.write(`[llm-dummy] received ${sig}; shutting down\n`);
    server.close(() => process.exit(0));
    setTimeout(() => process.exit(1), 5_000).unref();
  };
  process.on('SIGINT', () => shutdown('SIGINT'));
  process.on('SIGTERM', () => shutdown('SIGTERM'));
}

main();
